documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,799 @@
+         2. [5.8.3.2. Stage skipping](#5832-stage-skipping)
+         3. [5.8.3.3. Multi-stage chaining](#5833-multi-stage-chaining)
+      4. [5.8.4. Action publishing](#584-action-publishing)
+      5. [5.8.5. Metadata publishing](#585-metadata-publishing)
+   9. [5.9. Action apply](#59-action-apply)
+   10. [5.10. Meter update](#510-meter-update)
+6. [6. Examples](#6-examples)
+   1. [6.1. VNET routing](#61-vnet-routing)
+   2. [6.2. VM level public IP inbound (L3 DNAT)](#62-vm-level-public-ip-inbound-l3-dnat)
+   3. [6.3. VM level public IP outbound (L3 SNAT)](#63-vm-level-public-ip-outbound-l3-snat)
+   4. [6.4. Load balancer (L4 DNAT)](#64-load-balancer-l4-dnat)
+   5. [6.5. More](#65-more)
+
+## 1. Overview
+
+DASH-SAI pipeline packet flow is the core of the DASH project. It defines from device perspective, how the traffic is modeled in DASH, packets gets processed and transformation gets applied to the packets.
+
+DASH-SAI pipeline is designed to work as a general purpose network function pipeline. Similar to [SAI](https://github.com/opencomputeproject/SAI), it works as a shim layer on top of DPU/ASIC SDKs, provides a set of low level hardware agnostic APIs that exposes the generic DPU primitives to the upper layer. So, the pipeline itself is not limited to any specific network function, but can be used to implement any network function.
+
+## 2. Data path logical architecture stack
+
+[DASH HLD](https://github.com/sonic-net/DASH/blob/main/documentation/general/dash-high-level-design.md) already provided a very good high level overview on the system architecture. And here, we are going to dive a bit deeper from the data path perspective.
+
+![](./images/dash-data-path-overview.svg)
+
+DASH-SAI APIs are designed to be generic, so we could use it to set up the data path in different ways as needed. 
+
+- The simplest way to set up the data path is to use the DASH-SAI APIs to create your pipeline and offload all your policy to ASIC.
+- If your application requires more advanced policy, then the techonology provider could also provide a inbox data plane app to implement these policies. And still, this is transparent to the upper layer, everything is hidden under the SAI level.
+- If even more customized or complicated pipelines are needed for certain traffic, we could also provide our own customized data plane app and use the DASH policy to instruct the ASIC to trap these packets up when they shows up.
+
+## 3. Packet Structure
+
+Before diving into the pipeline and packet flow, to better describe the behaviors, let's first have a clear understanding on the packet structure that DASH supports.
+
+Overall, the high-level packet structure looks like below:
+
+- The inner most packet is whatever the customer sends, which is called overlay. 
+- The first encap is called underlay, which is the most frequently used layer for implement any virtual network functions, such as VNET routing, load balancer, etc.
```

Comment:
Since the underlay also requires a tunnel encapsulation, wouldn't it be simple to model the layers as | ** | Tunnel 2 | Tunnel 1 | Tunnel 0 | Overlay | where Tunnel0 is used for underlay?

Reply:
that's actually true. however, this underlay concepts is already everywhere in today's DASH pipeline, such as mapping table, eni and etc.... so maybe better to keep it as it is for now.

but to allow this change for future, we could use | ** | Tunnel 2 | Tunnel 1 | Underlay | Overlay |. does this sound good to you?

Reply:
updated.

Reply:
Yeah, that's much better. Now, Tunne0 and underlay can be used interchangeably, if needed.